src/python/pants/backend/jvm/tasks/reports/junit_html_report.py

```diff
@@ -5,23 +5,93 @@
+import itertools
+import logging
 import os
 import xml.etree.ElementTree as ET
 from abc import abstractmethod
 from functools import total_ordering
 
+from twitter.common.collections import OrderedSet
+
 from pants.base.mustache import MustacheRenderer
 from pants.util import desktop
-from pants.util.dirutil import safe_mkdir_for
+from pants.util.dirutil import safe_mkdir_for, safe_walk
+from pants.util.memo import memoized_property
 from pants.util.meta import AbstractClass
+from pants.util.objects import datatype
 from pants.util.strutil import ensure_binary
 
 
+_LOGGER = logging.getLogger(__name__)
+
+
 @total_ordering
 class ReportTestSuite(object):
   """Data object for a JUnit test suite"""
 
+  class MergeError(Exception):
+    def __init__(self, test_cases):
+      error_message = 'Conflicting test cases are not mergeable:\n\t{}'.format(
+        '\n\t'.join(map(repr, test_cases)))
+      super(ReportTestSuite.MergeError, self).__init__(error_message)
+
+  @classmethod
+  def merged(cls, report_test_suites, error_on_conflict=True, logger=None):
+    """Merges any like-named test suites into one.
+
+    :param report_test_suites: A sequence of test suites to merge results from.
+    :type report_test_suites: :class:`collections.Iterable` of :class:`ReportTestSuite`
+    :param bool error_on_conflict: `True` to raise when two or more test cases in a given test suite
+                                   have the same name but otherwise different attributes.
+    :param logger: An optional logger to use for logging merge conflicts.
+    :type logger: :class:`logging.Logger`
+    :raises: :class:`ReportTestSuite.MergeError` if configured to do so on merge errors.
+    :returns: A single test suite with the result of all the given `report_test_suites` merged.
+    :rtype: iter of :class:`ReportTestSuite`s
+    """
+
+    logger = logger or _LOGGER
+
+    suites_by_name = collections.defaultdict(list)
+    for report_test_suite in report_test_suites:
+      suites_by_name[report_test_suite.name].append(report_test_suite)
+
+    for name, suites in suites_by_name.items():
+      cases_by_name = collections.defaultdict(OrderedSet)
+      for case in itertools.chain.from_iterable(s.testcases for s in suites):
+        cases_by_name[case.name].add(case)
+
+      test_cases = []
+      tests, errors, failures, skipped, time = 0, 0, 0, 0, 0
+      for cases in cases_by_name.values():
+        if len(cases) > 1:
+          if error_on_conflict:
+            raise cls.MergeError(cases)
+          else:
+            logger.warning('Found test case results with conflicting values, using first result '
+                           'from:\n\t{}'.format('\n\t'.join(map(str, cases))))
+        case = iter(cases).next()
+        tests += 1
+        time += case.time
+        if case.error:
+          errors += 1
+        elif case.failure:
+          failures += 1
+        elif case.skipped:
+          skipped += 1
+        test_cases.append(case)
```

Comment:
Duplicates (conflicting or otherwise) are expected never to happen, but the merge logic does present them by raising or logging for logical completeness only. This is all just a sanity check that the user could do nothing about.